Further use C++ concepts in rest of WebCore

iidmsa opened the pull request:
#### 4fc477a8449a25e3549d211782c955c35bd86d1a
<pre>
Further use C++ concepts in rest of WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=294128">https://bugs.webkit.org/show_bug.cgi?id=294128</a>
<a href="https://rdar.apple.com/problem/152726842">rdar://problem/152726842</a>

Reviewed by Sam Weinig.

Replace std::enable_if with C++20 concepts and require clauses for template constraints.

* Source/WebCore/bindings/js/JSDOMConvertInterface.h:
(): Deleted.
* Source/WebCore/bindings/js/JSDOMIterator.h:
* Source/WebCore/bindings/js/WebCoreOpaqueRoot.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateImplementation):
* Source/WebCore/dom/ExceptionOr.h:
(WebCore::ExceptionOr&lt;ReturnType&gt;::ExceptionOr):
* Source/WebCore/dom/GCReachableRef.h:
(WTF::GetPtrHelper&lt;WebCore::GCReachableRef&lt;T&gt;&gt;::getPtr):

Canonical link: <a href="https://commits.webkit.org/296043@main">https://commits.webkit.org/296043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a5071bb15d5ce31e6f3d66b7e04bda18a49b634

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26496 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111956 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57343 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108784 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27166 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34998 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81046 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109749 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21555 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96323 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61387 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14430 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56797 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90895 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14459 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114883 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33883 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24978 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90111 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92554 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89821 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22980 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34758 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12569 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29530 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33808 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39221 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33554 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36907 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35153 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -5438,7 +5438,10 @@ extern "C" { extern void (*const ${vtableRefWin}[])(); }
-template<typename T, typename = std::enable_if_t<std::is_same_v<T, ${implType}>, void>> static inline void verifyVTable(${implType}* ptr) {
+template<typename T>
+    requires std::same_as<T, ${implType}>
+static inline void verifyVTable(${implType}* ptr) 
```
I think many of these same_as requirements can be done as 

```c++
template<std::same_as<${implType}> T>
static inline ....
```

iidmsa replied:
Fixed.

iidmsa committed:
Further use C++ concepts in rest of WebCore
https://bugs.webkit.org/show_bug.cgi?id=294128
rdar://problem/152726842

Reviewed by Sam Weinig.

Replace std::enable_if with C++20 concepts and require clauses for template constraints.

* Source/WebCore/bindings/js/JSDOMConvertInterface.h:
(): Deleted.
* Source/WebCore/bindings/js/JSDOMIterator.h:
* Source/WebCore/bindings/js/WebCoreOpaqueRoot.h:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateImplementation):
* Source/WebCore/dom/ExceptionOr.h:
(WebCore::ExceptionOr<ReturnType>::ExceptionOr):
* Source/WebCore/dom/GCReachableRef.h:
(WTF::GetPtrHelper<WebCore::GCReachableRef<T>>::getPtr):

Canonical link: https://commits.webkit.org/296043@main
